### PR TITLE
Vendor ethPM

### PR DESCRIPTION
**Pull request by njgheorghita**
### What was wrong?
Things that happened here (aka why is this pr so large)
- `py-ethpm` codebase was relocated to `web3/ethpm`
- critical components moved from `pytest-ethereum` to `web3.tools.pytest_ethereum`
- `pytest-ethereum` plugins exposed via `entrypoints` in `setup.py`
- removed `web3`'s core dependencies on `py-ethpm` and `pytest-ethereum`
- added a couple core dependencies necessary for `ethpm`
- a local ipfs node now runs in the background of circle builds
- `ethpm` documentation has been moved to `web3` docs

Things to note
- No code has been changed / needs review in the migrated files, so the general structure of the migrated files is what primarily needs review in this pr.

#### Cute Animal Picture
image

### CODE REVIEW

**Comment by voith** [comment]
@njgheorghita Out of curiosity, what is the reason behind relocating `ethpm` to `web3.py`?

**Comment by njgheorghita** [comment]
@voith Totally fair, it basically comes down to a circular dependency between `web3` and `ethpm` that's causing [some](https://github.com/ethpm/py-ethpm/issues/140) [issues](https://github.com/ethpm/py-ethpm/pull/137#issuecomment-449873953). `web3` depends on `ethpm` concepts in `web3.pm` and `ethpm` uses `web3` concepts liberally. My impression is the best solution to remove this dependency is to vendor `ethpm` and all relevant `pytest-ethereum` code in `web3`. It adds a fair amount of code overhead to the `web3` codebase, but the goal is to keep it well separated from the `web3` core code.

**Comment by njgheorghita** [comment]
@kclowes Yup! Exactly, `pytest-ethereum` has the same circular dependency problem with `web3` that `ethpm` had. However, since `pytest-ethereum` doesn't rely on `web3` as heavily as `ethpm`, only the few core components were migrated into the `web3` codebase rather than the entire library. So, `pytest-ethereum` will continue to exist as a wrapper around and extensions beyond these new `web3` components.

File: setup.py [hunk 1]
```diff
@@ -9,7 +9,6 @@
     'tester': [
         "eth-tester[py-evm]==0.1.0-beta.39",
         "py-geth>=2.0.1,<3.0.0",
-        "pytest-ethereum>=0.1.3a6,<1.0.0",
```
Comment (carver) [inline setup.py]: Sweet

File: docs/ethpm.rst [hunk 2]
```diff
@@ -0,0 +1,854 @@
...
+        ...
+   ethpm.exceptions.BytecodeLinkingError: Contract cannot be deployed until its bytecode is linked.
+
+   >>> # Deploy SafeSendLib
+   >>> SafeSendFactory = EscrowPackage.get_contract_factory("SafeSendLib")
+   >>> safe_send_tx_hash = SafeSendFactory.constructor().transact()
+   >>> safe_send_tx_receipt = w3.eth.waitForTransactionReceipt(safe_send_tx_hash)
+
+   >>> # Link Escrow factory to deployed SafeSendLib instance
+   >>> LinkedEscrowFactory = EscrowFactory.link_bytecode({"SafeSendLib": to_canonical_address(safe_send_tx_receipt.contractAddress)})
```
[the remark this answers is not in the document]
Comment (njgheorghita) [reply docs/ethpm.rst]: Yup, that makes sense - will take care of this in a following pr to keep review of this pr relatively simple

**Pull request merged**